# Netmiko 

Netmiko is a python library used to connect to device/file transfer. 

Developed based on paramiko and textfsm.

Using Juniper devices for example

Netmiko inherits JuniperSSH,JuniperTelnet
->JuniperBase ##give all the available function for ConnectionClass.
->Baseconnection ##send_command is defined here


# conncet to device

In [4]:
from netmiko import ConnectHandler

juniper_mx960={
    'device_type': 'juniper_junos',
    'host':   '10.85.160.9',
    'username': 'labroot',
    'password': 'lab123',
}

mx960_connection=ConnectHandler(**juniper_mx960)
output_qfx=mx960_connection.send_command('show version')
print(output_qfx)

Hostname: grapevine-re0
Model: mx960
Junos: 19.3R3.6
JUNOS OS Kernel 64-bit  [20200710.a76c78d_builder_stable_11]
JUNOS OS libs [20200710.a76c78d_builder_stable_11]
JUNOS OS runtime [20200710.a76c78d_builder_stable_11]
JUNOS OS time zone information [20200710.a76c78d_builder_stable_11]
JUNOS network stack and utilities [20200828.051203_builder_junos_193_r3]
JUNOS libs [20200828.051203_builder_junos_193_r3]
JUNOS OS libs compat32 [20200710.a76c78d_builder_stable_11]
JUNOS OS 32-bit compatibility [20200710.a76c78d_builder_stable_11]
JUNOS libs compat32 [20200828.051203_builder_junos_193_r3]
JUNOS runtime [20200828.051203_builder_junos_193_r3]
JUNOS sflow mx [20200828.051203_builder_junos_193_r3]
JUNOS py extensions [20200828.051203_builder_junos_193_r3]
JUNOS py base [20200828.051203_builder_junos_193_r3]
JUNOS OS crypto [20200710.a76c78d_builder_stable_11]
JUNOS na telemetry [19.3R3.6]
JUNOS Web Management [20200828.051203_builder_junos_193_r3]
JUNOS Security Intelligence [20200828.0512

# Check all available device type:
    
    

In [8]:
#1.Could check ssh.dispatcher under netmiko directory

#2. Give an invalid device type in the code. 

from netmiko import ConnectHandler

try:
    juniper_mx960={
        'device_type': 'checking',
        'host':   '10.85.160.9',
        'username': 'labroot',
        'password': 'lab123',
    }

    mx960_connection=ConnectHandler(**juniper_mx960)
    
except ValueError as e:
    print(e)

Unsupported 'device_type' currently supported platforms are: 
a10
accedian
adtran_os
alcatel_aos
alcatel_sros
apresia_aeos
arista_eos
aruba_os
aruba_osswitch
aruba_procurve
avaya_ers
avaya_vsp
broadcom_icos
brocade_fastiron
brocade_netiron
brocade_nos
brocade_vdx
brocade_vyos
calix_b6
centec_os
checkpoint_gaia
ciena_saos
cisco_asa
cisco_ftd
cisco_ios
cisco_nxos
cisco_s300
cisco_tp
cisco_wlc
cisco_xe
cisco_xr
cloudgenix_ion
coriant
dell_dnos9
dell_force10
dell_isilon
dell_os10
dell_os6
dell_os9
dell_powerconnect
dlink_ds
eltex
eltex_esr
endace
enterasys
extreme
extreme_ers
extreme_exos
extreme_netiron
extreme_nos
extreme_slx
extreme_vdx
extreme_vsp
extreme_wing
f5_linux
f5_ltm
f5_tmsh
flexvnf
fortinet
generic
generic_termserver
hp_comware
hp_procurve
huawei
huawei_olt
huawei_smartax
huawei_vrpv8
ipinfusion_ocnos
juniper
juniper_junos
juniper_screenos
keymile
keymile_nos
linux
mellanox
mellanox_mlnxos
mikrotik_routeros
mikrotik_switchos
mrv_lx
mrv_optiswitch
netapp_cdot
netgear_prosafe
n

# Cool feature, session log


In [15]:
from netmiko import ConnectHandler

juniper_mx960={
    'device_type': 'juniper_junos',
    'host':   '10.85.160.9',
    'username': 'labroot',
    'password': 'lab123',
    'session_log':'output.txt'
}

mx960_connection=ConnectHandler(**juniper_mx960)
output_qfx=mx960_connection.send_command('show version')
print(output_qfx)

Hostname: grapevine-re0
Model: mx960
Junos: 19.3R3.6
JUNOS OS Kernel 64-bit  [20200710.a76c78d_builder_stable_11]
JUNOS OS libs [20200710.a76c78d_builder_stable_11]
JUNOS OS runtime [20200710.a76c78d_builder_stable_11]
JUNOS OS time zone information [20200710.a76c78d_builder_stable_11]
JUNOS network stack and utilities [20200828.051203_builder_junos_193_r3]
JUNOS libs [20200828.051203_builder_junos_193_r3]
JUNOS OS libs compat32 [20200710.a76c78d_builder_stable_11]
JUNOS OS 32-bit compatibility [20200710.a76c78d_builder_stable_11]
JUNOS libs compat32 [20200828.051203_builder_junos_193_r3]
JUNOS runtime [20200828.051203_builder_junos_193_r3]
JUNOS sflow mx [20200828.051203_builder_junos_193_r3]
JUNOS py extensions [20200828.051203_builder_junos_193_r3]
JUNOS py base [20200828.051203_builder_junos_193_r3]
JUNOS OS crypto [20200710.a76c78d_builder_stable_11]
JUNOS na telemetry [19.3R3.6]
JUNOS Web Management [20200828.051203_builder_junos_193_r3]
JUNOS Security Intelligence [20200828.0512

from output.txt file:
    
Last login: Mon Dec 21 13:27:51 2020 from 10.110.23.187
--- JUNOS 19.3R3.6 Kernel 64-bit  JNPR-11.0-20200710.a76c78d_buil
labroot@grapevine-re0> 

labroot@grapevine-re0> 

labroot@grapevine-re0> cli
                       ^
unknown command.

labroot@grapevine-re0> 

labroot@grapevine-re0> 

labroot@grapevine-re0> set cli complete-on-space off 
Disabling complete-on-space

labroot@grapevine-re0> set cli screen-width 511 
Screen width set to 511

labroot@grapevine-re0> set cli screen-length 0 
Screen length set to 0

labroot@grapevine-re0> 

labroot@grapevine-re0> 

labroot@grapevine-re0> show version 
Hostname: grapevine-re0
Model: mx960
Junos: 19.3R3.6
JUNOS OS Kernel 64-bit  [20200710.a76c78d_builder_stable_11]
JUNOS OS libs [20200710.a76c78d_builder_stable_11]
JUNOS OS runtime [20200710.a76c78d_builder_stable_11]
JUNOS OS time zone information [20200710.a76c78d_builder_stable_11]
JUNOS network stack and utilities [20200828.051203_builder_junos_193_r3]
JUNOS libs [20200828.051203_builder_junos_193_r3]
JUNOS OS libs compat32 [20200710.a76c78d_builder_stable_11]
JUNOS OS 32-bit compatibility [20200710.a76c78d_builder_stable_11]
JUNOS libs compat32 [20200828.051203_builder_junos_193_r3]
JUNOS runtime [20200828.051203_builder_junos_193_r3]
JUNOS sflow mx [20200828.051203_builder_junos_193_r3]
JUNOS py extensions [20200828.051203_builder_junos_193_r3]
JUNOS py base [20200828.051203_builder_junos_193_r3]
JUNOS OS crypto [20200710.a76c78d_builder_stable_11]
JUNOS na telemetry [19.3R3.6]
JUNOS Web Management [20200828.051203_builder_junos_193_r3]
JUNOS Security Intelligence [20200828.051203_builder_junos_193_r3]
JUNOS mx libs compat32 [20200828.051203_builder_junos_193_r3]
JUNOS mx runtime [20200828.051203_builder_junos_193_r3]
JUNOS RPD Telemetry Application [19.3R3.6]
JUNOS common platform support [20200828.051203_builder_junos_193_r3]
JUNOS Openconfig [19.3R3.6]
JUNOS mtx network modules [20200828.051203_builder_junos_193_r3]
JUNOS modules [20200828.051203_builder_junos_193_r3]
JUNOS mx modules [20200828.051203_builder_junos_193_r3]
JUNOS mx libs [20200828.051203_builder_junos_193_r3]
JUNOS SQL Sync Daemon [20200828.051203_builder_junos_193_r3]
JUNOS mtx Data Plane Crypto Support [20200828.051203_builder_junos_193_r3]
JUNOS daemons [20200828.051203_builder_junos_193_r3]
JUNOS mx daemons [20200828.051203_builder_junos_193_r3]
JUNOS appidd-mx application-identification daemon [20200828.051203_builder_junos_193_r3]
JUNOS Services URL Filter package [20200828.051203_builder_junos_193_r3]
JUNOS Services TLB Service PIC package [20200828.051203_builder_junos_193_r3]
JUNOS Services Telemetry [20200828.051203_builder_junos_193_r3]
JUNOS Services TCP-LOG [20200828.051203_builder_junos_193_r3]
JUNOS Services SSL [20200828.051203_builder_junos_193_r3]
JUNOS Services SOFTWIRE [20200828.051203_builder_junos_193_r3]
JUNOS Services Stateful Firewall [20200828.051203_builder_junos_193_r3]
JUNOS Services RTCOM [20200828.051203_builder_junos_193_r3]
JUNOS Services RPM [20200828.051203_builder_junos_193_r3]
JUNOS Services PCEF package [20200828.051203_builder_junos_193_r3]
JUNOS Services NAT [20200828.051203_builder_junos_193_r3]
JUNOS Services Mobile Subscriber Service Container package [20200828.051203_builder_junos_193_r3]
JUNOS Services MobileNext Software package [20200828.051203_builder_junos_193_r3]
JUNOS Services Logging Report Framework package [20200828.051203_builder_junos_193_r3]
JUNOS Services LL-PDF Container package [20200828.051203_builder_junos_193_r3]
JUNOS Services Jflow Container package [20200828.051203_builder_junos_193_r3]
JUNOS Services Deep Packet Inspection package [20200828.051203_builder_junos_193_r3]
JUNOS Services IPSec [20200828.051203_builder_junos_193_r3]
JUNOS Services IDS [20200828.051203_builder_junos_193_r3]
JUNOS IDP Services [20200828.051203_builder_junos_193_r3]
JUNOS Services HTTP Content Management package [20200828.051203_builder_junos_193_r3]
JUNOS Services Crypto [20200828.051203_builder_junos_193_r3]
JUNOS Services Captive Portal and Content Delivery Container package [20200828.051203_builder_junos_193_r3]
JUNOS Services COS [20200828.051203_builder_junos_193_r3]
JUNOS AppId Services [20200828.051203_builder_junos_193_r3]
JUNOS Services Application Level Gateways [20200828.051203_builder_junos_193_r3]
JUNOS Services AACL Container package [20200828.051203_builder_junos_193_r3]
JUNOS SDN Software Suite [20200828.051203_builder_junos_193_r3]
JUNOS Extension Toolkit [20200828.051203_builder_junos_193_r3]
JUNOS Packet Forwarding Engine Support (wrlinux9) [20200828.051203_builder_junos_193_r3]
JUNOS Packet Forwarding Engine Support (MXSPC3) [19.3R3.6]
JUNOS Packet Forwarding Engine Support (MX/EX92XX Common) [20200828.051203_builder_junos_193_r3]
JUNOS Packet Forwarding Engine Support (M/T Common) [20200828.051203_builder_junos_193_r3]
JUNOS Packet Forwarding Engine Support (aft) [20200828.051203_builder_junos_193_r3]
JUNOS Packet Forwarding Engine Support (MX Common) [20200828.051203_builder_junos_193_r3]
JUNOS Juniper Malware Removal Tool (JMRT) [1.0.0+20200828.051203_builder_junos_193_r3]
JUNOS J-Insight [20200828.051203_builder_junos_193_r3]
JUNOS jfirmware [20200828.051203_builder_junos_193_r3]
JUNOS Online Documentation [20200828.051203_builder_junos_193_r3]
JUNOS jail runtime [20200710.a76c78d_builder_stable_11]
JUNOS FIPS mode utilities [20200828.051203_builder_junos_193_r3]

labroot@grapevine-re0> 

# Use TextFSM

In [12]:
from netmiko import ConnectHandler
from pprint import pprint

juniper_mx960={
    'device_type': 'juniper_junos',
    'host':   '10.85.160.9',
    'username': 'labroot',
    'password': 'lab123',
}

mx960_connection=ConnectHandler(**juniper_mx960)
output_qfx=mx960_connection.send_command('show version',use_textfsm=True)
pprint(output_qfx)

[{'appid_services': '20200828.051203_builder_junos_193_r3',
  'base_os_boot': '',
  'base_os_software_suite': '',
  'border_gateway_function_package': '',
  'crypto_software_suite': '',
  'hostname': 'grapevine-re0',
  'idp_services': '20200828.051203_builder_junos_193_r3',
  'junos_version': '19.3R3.6',
  'kernel_software_suite': '',
  'lab_package': '',
  'model': 'mx960',
  'online_documentation': '20200828.051203_builder_junos_193_r3',
  'other_device_properties': ['OS Kernel 64-bit  ',
                              'OS libs ',
                              'OS runtime ',
                              'OS time zone information ',
                              'network stack and utilities ',
                              'libs ',
                              'OS libs compat32 ',
                              'OS 32-bit compatibility ',
                              'libs compat32 ',
                              'runtime ',
                              'sflow mx ',
               

Q: How does Netmiko map the command output with specific textfsm temaplate?

NTC template: https://github.com/networktocode/ntc-templates

NTC template has pre-defined template for multiple vendors

Mapping is done via index file under the template dir:

Template, Hostname, Platform, Command

juniper_junos_show_version.textfsm, .*, juniper_junos, sh[[ow]] ver[[sion]]
juniper_junos_show_chassis_cluster_interfaces.textfsm, .*, juniper_junos, sh[[ow]] ch[[assis]] c[[luster]] i[[nterface]]
juniper_junos_show_ethernet-switching_table.textfsm, .*, juniper_junos, sh[[ow]] et[[hernet-switching]] t[[able]]
juniper_junos_show_chassis_cluster_status.textfsm, .*, juniper_junos, sh[[ow]] ch[[assis]] c[[luster]] s[[tatus]]
juniper_junos_show_chassis_firmware.textfsm, .*, juniper_junos, sh[[ow]] ch[[assis]] fi[[rmware]]
juniper_junos_show_lacp_interfaces.textfsm, .*, juniper_junos, sh[[ow]] la[[cp]] i[[nterfaces]]
juniper_junos_show_arp_no-resolve.textfsm, .*, juniper_junos, sh[[ow]] a[[rp]] n[[o-resolve]]
juniper_junos_show_isis_adjacency.textfsm, .*, juniper_junos, sh[[ow]] is[[is]] ad[[jacency]]
juniper_junos_show_ospf_neighbor.textfsm, .*, juniper_junos, sh[[ow]] ospf n[[eighbor]]
juniper_junos_show_interfaces.textfsm, .*, juniper_junos, sh[[ow]] inte[[rfaces]]


From the source code of send_command() in BaseConnection:

if use_textfsm:
            structured_output = get_structured_data(
                output,
                platform=self.device_type,
                command=command_string.strip(),
                template=textfsm_template,
            )
            # If we have structured data; return it.
            if not isinstance(structured_output, str):
                return structured_output


def get_structured_data(raw_output, platform=None, command=None, template=None):
    """
    Convert raw CLI output to structured data using TextFSM template.
    You can use a straight TextFSM file i.e. specify "template". If no template is specified,
    then you must use an CliTable index file.
    """
    if platform is None or command is None:
        attrs = {}
    else:
        attrs = {"Command": command, "Platform": platform}

    if template is None:
        if attrs == {}:
            raise ValueError(
                "Either 'platform/command' or 'template' must be specified."
            )
        template_dir = get_template_dir()
        index_file = os.path.join(template_dir, "index")
        textfsm_obj = clitable.CliTable(index_file, template_dir)
        return _textfsm_parse(textfsm_obj, raw_output, attrs)


It is done via Textfsm class CliTable:

Mapping the vendor name, command, hostname, and template together.


# Use your own template instead of the pre-defined template:

In [13]:
from netmiko import ConnectHandler


juniper_mx960={
    'device_type': 'juniper_junos',
    'host':   '10.85.160.9',
    'username': 'labroot',
    'password': 'lab123',
}



mx960_connection=ConnectHandler(**juniper_mx960)


output=mx960_connection.send_command('show version',use_textfsm=True,textfsm_template='customized_show_version.template')
print(output)

[{'hostname': 'grapevine-re0', 'model': 'mx960', 'junos_version': '19.3R3.6', 'base_os_boot': '', 'base_os_software_suite': '', 'kernel_software_suite': '', 'crypto_software_suite': '', 'packet_forwarding_engine_support_m_t_ex_common': '', 'packet_forwarding_engine_support_mx_common': '20200828.051203_builder_junos_193_r3', 'online_documentation': '20200828.051203_builder_junos_193_r3', 'services_aacl_container_package': '20200828.051203_builder_junos_193_r3', 'services_application_level_gateways': '20200828.051203_builder_junos_193_r3', 'appid_services': '20200828.051203_builder_junos_193_r3', 'border_gateway_function_package': '', 'services_captive_portal_content_delivery_package': '20200828.051203_builder_junos_193_r3', 'services_http_content_management_package': '20200828.051203_builder_junos_193_r3', 'idp_services': '20200828.051203_builder_junos_193_r3', 'services_jflow_container_package': '20200828.051203_builder_junos_193_r3', 'services_ll_pdf_container_package': '20200828.0512

# Create your own Template,Vendor,command and hostname mapping

In [14]:
from textfsm import clitable
from netmiko import ConnectHandler
from pprint import pprint



juniper_qfx10k2={
    'device_type': 'juniper_junos',
    'host':   '10.85.168.43',
    'username': 'labroot',
    'password': 'lab123',

}

qfx_qfx10k2_connection=netmiko.ConnectHandler(**juniper_qfx10k2)
qfx_show_version=qfx_qfx10k2_connection.send_command('show version')

print(qfx_show_version)

cli_table=clitable.CliTable('index','.')
attr={'Hostname': 'qfx10k2', 'Command': 'show version' , 'Vendor': 'Juniper'}
cli_table.ParseCmd(qfx_show_version,attr)
print(cli_table)


Hostname: jtac-QFX10002-72Q-r014
Model: qfx10002-72q
Junos: 18.4R2-S3 flex
JUNOS OS Kernel 64-bit FLEX [20191223.5f5c7dc_builder_stable_11]
JUNOS OS libs [20191223.5f5c7dc_builder_stable_11]
JUNOS OS runtime [20191223.5f5c7dc_builder_stable_11]
JUNOS OS time zone information [20191223.5f5c7dc_builder_stable_11]
JUNOS OS libs compat32 [20191223.5f5c7dc_builder_stable_11]
JUNOS OS 32-bit compatibility [20191223.5f5c7dc_builder_stable_11]
JUNOS py extensions [20191226.165110_builder_junos_184_r2_s3]
JUNOS py base [20191226.165110_builder_junos_184_r2_s3]
JUNOS OS vmguest [20191223.5f5c7dc_builder_stable_11]
JUNOS OS crypto [20191223.5f5c7dc_builder_stable_11]
JUNOS network stack and utilities [20191226.165110_builder_junos_184_r2_s3]
JUNOS libs [20191226.165110_builder_junos_184_r2_s3]
JUNOS libs compat32 [20191226.165110_builder_junos_184_r2_s3]
JUNOS runtime [20191226.165110_builder_junos_184_r2_s3]
JUNOS na telemetry [18.4R2-S3]
JUNOS Web Management Platform Package [20191226.165110_bu

# Use Netmiko to do file transfer

In [16]:
from netmiko import ConnectHandler,file_transfer

juniper_qfx10k2={
    'device_type': 'juniper_junos',
    'host':   '10.85.168.43',
    'username': 'labroot',
    'password': 'lab123',

}

qfx_qfx10k2_connection=netmiko.ConnectHandler(**juniper_qfx10k2)

transfer_result=file_transfer(qfx_qfx10k2_connection,source_file='output.txt',dest_file='output.txt',file_system='/var/tmp',direction='put',overwrite_file=True)
print(transfer_result)

{'file_exists': True, 'file_transferred': True, 'file_verified': True}


labroot@jtac-QFX10002-72Q-r014> file list /var/tmp | match output 
output.txt